### PR TITLE
removing graph from inputs, adding checksums

### PR DESCRIPTION
**Pull request by djarecka**
## Checklist

- [ x] All tests passing
- [ x] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [ x] My code follows the code style of this project 
(we are using `black`: you can `pip install pre-commit`, 
run `pre-commit install` in the `pydra` directory 
and `black` will be run automatically with each commit)

## Summary

saving checksums of nodes from the graph instead of copy of entire graph in `wf.inputs` (we only need it for calculating checksum anyway)

## Acknowledgment
- [x ] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

**Comment by djarecka** [comment]
this assumes (as in master) that we're not changing nodes after adding it to wf

File: pydra/engine/core.py [hunk 1]
```diff
@@ -508,7 +508,7 @@ def add(self, task):
         self._last_added = task
         self.create_connections(task)
         logger.debug(f"Added {task}")
-        self.inputs._graph = deepcopy(self.graph_sorted)
+        self.inputs._graph_checksums = [nd.checksum for nd in self.graph_sorted]
```
Comment (satra) [inline pydra/engine/core.py]: this would compute graph and node checksum multiple times for each add.

File: pydra/engine/specs.py [hunk 2]
```diff
@@ -25,13 +25,11 @@ def hash(self):
...
-            if field.name not in ["_graph"]
+            if field.name not in ["_graph_checksums"]
         }
         inp_hash = hash_function(inp_dict)
-        if hasattr(self, "_graph"):
-            # calculating checksum for every node inside the graph
-            graph_hash = [nd.checksum for nd in self._graph]
-            return hash_function((inp_hash, graph_hash))
+        if hasattr(self, "_graph_checksums"):
+            return hash_function((inp_hash, self._graph_checksums))
```
Comment (satra) [inline pydra/engine/specs.py]: i would compute the graph checksum at this step.

File: pydra/engine/core.py (hunk 1 above)
Comment (djarecka) [reply pydra/engine/core.py]: the graph has to be recomputed (it's sorted)

File: pydra/engine/specs.py (hunk 2 above)
Comment (djarecka) [reply pydra/engine/specs.py]: but that was in the current master...

Comment (satra) [reply pydra/engine/specs.py]: in the current master `graph_hash = [nd.checksum for nd in self._graph]` this computes the hash for every node at the point of checksum computation, not in the `add` function.

File: pydra/engine/core.py (hunk 1 above)
Comment (satra) [reply pydra/engine/core.py]: all i am saying is compute the checksum only when it is needed.

File: pydra/engine/specs.py (hunk 2 above)
Comment (djarecka) [reply pydra/engine/specs.py]: in that case I don't know how to remove deepcopy. Nodes are changing, so if we don't create copy and we calculate whenever we want, than the checksum will be changing during execution, and we don't want it.

**Comment by satra** [comment]
i don't think we want to deepcopy in the `add` function, but when the graph checksum is being computed.

**Comment by satra** [comment]
is this now ready?

**Comment by djarecka** [comment]
@satra - it might be good if you review the last changes. If you like it, you can merge

File: pydra/engine/tests/test_workflow.py [hunk 3]
```diff
@@ -1438,6 +1632,68 @@ def test_wf_ndstate_cachelocations(plugin, tmpdir):
...
+    with Submitter(plugin=plugin) as sub:
+        sub(wf2)
+    t2 = time.time() - t0
+
+    results2 = wf2.result()
+    assert results2.output.out == [8, 82]
+
+    # checking execution time
+    assert t1 > 3
+    assert t2 < 0.1
```
Comment (satra) [inline pydra/engine/tests/test_workflow.py]: i'm going to merge this, but the way i think we should evaluate whether the correct tasks are rerun is using messaging. (see #9)

Comment (djarecka) [reply pydra/engine/tests/test_workflow.py]: are you sure that you pointed me to the right pr?

Comment (satra) [reply pydra/engine/tests/test_workflow.py]: oops - updated the link

Comment (djarecka) [reply pydra/engine/tests/test_workflow.py]: yes, we should add it, but just fyi, the time is not the only check in the tests, I'm also checking if the outputdir is created.

**Pull request merged**